### PR TITLE
459 implement the function add the modules in neemc automatically and add the sipm readout in the module

### PR DESCRIPTION
**Pull request by johnny8266**
### Briefly, what does this PR introduce?
Add the automatic function to place the module.
Add the basic geometry of SiPM

### What kind of change does this PR introduce?
- [ ] Bug fix (issue #__)
- [X] New feature (issue #__)
- [ ] Documentation update
- [ ] Other: __

### Please check if this PR fulfills the following:
- [ ] Tests for the changes have been added
- [ ] Documentation has been added / updated
- [X] Changes have been communicated to collaborators

### Does this PR introduce breaking changes? What changes might users need to make to their code?
No
### Does this PR change default behavior?
No

### CODE REVIEW

File: src/HomogeneousCalorimeter_geo.cpp [hunk 1]
```diff
@@ -288,38 +303,67 @@ static std::tuple<int, std::pair<int, int>> add_12surface_disk(Detector& desc, A
...
+    }
 
-    Transform3D tr_local = RotationZYX(Nrot, 0.0, 0.0) * Translation3D(p.x(), p.y(), 0.0);
-    auto modPV = (has_envelope ? env_vol.placeVolume(modVol, tr_local) : env.placeVolume(modVol, tr_global * tr_local));
-    modPV.addPhysVolID("sector", sector_id).addPhysVolID("row", row).addPhysVolID("column", column);
   }
 
+  std::cout << "New Min: " << minX << ", " << minY << " || Max: " << maxX << ", " << maxY << std::endl;
+  std::cout << "Total count new: " << total_count << std::endl;
+
```
Comment (wdconinc) [inline src/HomogeneousCalorimeter_geo.cpp:366]: No output by default. Use `Printout` to `DEBUG`, e.g. https://github.com/eic/epic/blob/62f3a3803a9b95b9e367adc1f6391ef43ee98bfb/src/MPGDDIRC_geo.cpp#L210, then you can choose to run with log level set to debug.

File: src/HomogeneousCalorimeter_geo.cpp [hunk 2]
```diff
@@ -288,38 +303,67 @@ static std::tuple<int, std::pair<int, int>> add_12surface_disk(Detector& desc, A
...
 
-  for (auto& p : points) {
-    column = std::round((p.x() - minl_ptsx[0].x()) / modSize.x());
-    row = std::round((maxl_ptsy[0].y() - p.y()) / modSize.y());
+  for (auto &square : test_points) {
+    epic::geo::Point box[4] = {{square.x() + half_modx, square.y() + half_mody}, {square.x() - half_modx, square.y() + half_mody}, {square.x() - half_modx, square.y() - half_mody}, {square.x() + half_modx, square.y() - half_mody}};
+
+    if (epic::geo::isBoxInsidePolygon(box, Out_vertices)) {
+      if(!epic::geo::isBoxInsidePolygon(box, In_vertices)) {
+        // std::cout << "Square inside the inner frame: " << square.x() << ", " << square.y() << std::endl;
```
Comment (wdconinc) [inline src/HomogeneousCalorimeter_geo.cpp:352]: No commented out debug lines.

File: src/HomogeneousCalorimeter_geo.cpp [hunk 3]
```diff
@@ -288,38 +303,67 @@ static std::tuple<int, std::pair<int, int>> add_12surface_disk(Detector& desc, A
...
+  auto polyVertex = epic::geo::getPolygonVertices(c1, (rmax/std::cos(Prot)), M_PI/12., 12);
+  std::vector<epic::geo::Point> Out_vertices, In_vertices;
+  int index = 0;
+  for( auto p : polyVertex ){
+    epic::geo::Point a = {p.first, p.second};
+    Out_vertices.push_back(a);
+    // std::cout << "{" << Out_vertices[index].x() << ", " << Out_vertices[index].y() << "}, " << std::endl;
+    index++;
+  }
+  // std::cout << std::endl << std::endl;
```
Comment (wdconinc) [inline src/HomogeneousCalorimeter_geo.cpp:322]: No commented out debug lines.

File: src/HomogeneousCalorimeter_geo.cpp [hunk 4]
```diff
@@ -288,38 +303,67 @@ static std::tuple<int, std::pair<int, int>> add_12surface_disk(Detector& desc, A
...
+  auto test_points = epic::geo::fillRectangles({half_modx, half_mody}, modSize.x(), modSize.y(), 0., (rmax/std::cos(Prot)), phimin, phimax);
+
+  std::pair<double, double> c1 (0., 0.);
+  auto polyVertex = epic::geo::getPolygonVertices(c1, (rmax/std::cos(Prot)), M_PI/12., 12);
+  std::vector<epic::geo::Point> Out_vertices, In_vertices;
+  int index = 0;
+  for( auto p : polyVertex ){
+    epic::geo::Point a = {p.first, p.second};
+    Out_vertices.push_back(a);
+    // std::cout << "{" << Out_vertices[index].x() << ", " << Out_vertices[index].y() << "}, " << std::endl;
```
Comment (wdconinc) [inline src/HomogeneousCalorimeter_geo.cpp:319]: No commented out debug lines.

File: src/HomogeneousCalorimeter_geo.cpp [hunk 5]
```diff
@@ -288,38 +303,67 @@ static std::tuple<int, std::pair<int, int>> add_12surface_disk(Detector& desc, A
...
   //=====================================================================
 
-  // retrieve the max and min position of modules
-  auto [minl_ptsx, maxl_ptsx] = std::minmax_element(points.begin(), points.end(), [](auto a, auto b){ return a.x() < b.x(); });
-  auto [minl_ptsy, maxl_ptsy] = std::minmax_element(points.begin(), points.end(), [](auto a, auto b){ return a.y() < b.y(); });
+  auto test_points = epic::geo::fillRectangles({half_modx, half_mody}, modSize.x(), modSize.y(), 0., (rmax/std::cos(Prot)), phimin, phimax);
+
+  std::pair<double, double> c1 (0., 0.);
+  auto polyVertex = epic::geo::getPolygonVertices(c1, (rmax/std::cos(Prot)), M_PI/12., 12);
+  std::vector<epic::geo::Point> Out_vertices, In_vertices;
```
Comment (wdconinc) [inline src/HomogeneousCalorimeter_geo.cpp:314]: Local variables lowercase.
```suggestion
  std::vector<epic::geo::Point> out_vertices, in_vertices;
```

File: src/HomogeneousCalorimeter_geo.cpp (hunk 4 above)
Comment (johnny8266) [reply src/HomogeneousCalorimeter_geo.cpp]: remove in the last commit

File: src/HomogeneousCalorimeter_geo.cpp (hunk 5 above)
Comment (johnny8266) [reply src/HomogeneousCalorimeter_geo.cpp]: changed in the last commit

File: src/HomogeneousCalorimeter_geo.cpp (hunk 1 above)
Comment (johnny8266) [reply src/HomogeneousCalorimeter_geo.cpp]: Got it. change to the printout in the last commit

File: src/HomogeneousCalorimeter_geo.cpp (hunk 3 above)
Comment (johnny8266) [reply src/HomogeneousCalorimeter_geo.cpp]: remove it in the last commit

File: src/HomogeneousCalorimeter_geo.cpp (hunk 2 above)
Comment (johnny8266) [reply src/HomogeneousCalorimeter_geo.cpp]: remove it in the last commit

File: src/GeometryHelpers.cpp [hunk 6]
```diff
@@ -151,4 +151,86 @@ namespace epic::geo {
...
 
+
+  bool isPointInsidePolygon(Point p, std::vector<Point> vertices)
+  {
+    int n = vertices.size();
+    int i, j = 0;
+    bool check = false;  // check == false (outside the polygon), check == true (inside the polygon)
+    const double tolerance = 0.000001;
+
+    // When the point overlaped with vertex in the tolerance.
```
Comment (wdconinc) [inline src/GeometryHelpers.cpp:162]:
```suggestion
    // When the point overlaps with vertex in the tolerance.
```

File: src/GeometryHelpers.cpp [hunk 7]
```diff
@@ -151,4 +151,86 @@ namespace epic::geo {
     return res;
   }
 
+
+  bool isPointInsidePolygon(Point p, std::vector<Point> vertices)
+  {
+    int n = vertices.size();
+    int i, j = 0;
```
Comment (wdconinc) [inline src/GeometryHelpers.cpp:158]: Move `i` and `j` to local scope.

**Review by wdconinc — DISMISSED** [review]

File: src/GeometryHelpers.cpp [hunk 8]
```diff
@@ -151,4 +151,85 @@ namespace epic::geo {
...
+
+  bool isPointInsidePolygon(Point p, std::vector<Point> vertices)
+  {
+    int n = vertices.size();
+    bool check = false;  // check == false (outside the polygon), check == true (inside the polygon)
+    const double tolerance = 0.000001;
+
+    // When the point overlaps with vertex in the tolerance.
+    //
+    for( int i = 0, j = n-1 ; i < n ; j = i++)
```
Comment (wdconinc) [inline src/GeometryHelpers.cpp:163]:
```suggestion
    for( int i = 0 ; i < n ; i++)
```

**Review by wdconinc — APPROVED** [review]

**Pull request merged**